File: backend/src/bond_monitor/domain/portfolio/duration_metrics.py
"""Weighted duration metrics for portfolio plan and trading advice."""

from __future__ import annotations

from collections.abc import Sequence
from typing import Protocol

from bond_monitor.domain.bonds.models import BondRecord
from bond_monitor.domain.portfolio.models import PortfolioPosition


class _DurationHolding(Protocol):
    isin: str
    market_value_rub: float | None


def weighted_duration_by_purchase(
    positions: Sequence[PortfolioPosition],
    universe_by_isin: dict[str, BondRecord],
) -> float | None:
    """Средневзвешенная дюрация (годы), взвешенная по ``purchase_amount_rub``."""
    weight_total = 0.0
    weighted_sum = 0.0
    for position in positions:
        bond = universe_by_isin.get(position.isin)
        if bond is None or bond.duration_years is None:
            continue
        weight = position.purchase_amount_rub
        weight_total += weight
        weighted_sum += weight * bond.duration_years
    if weight_total <= 0:
        return None
    return weighted_sum / weight_total


def weighted_duration_by_market(
    holdings: Sequence[_DurationHolding],
    universe_by_isin: dict[str, BondRecord],
) -> float | None:
    """Средневзвешенная дюрация (годы), взвешенная по ``market_value_rub``."""
    weight_total = 0.0
    weighted_sum = 0.0
    for holding in holdings:
        bond = universe_by_isin.get(holding.isin)
        if bond is None or bond.duration_years is None:
            continue
        weight = holding.market_value_rub
        if weight is None or weight <= 0:
            continue
        weight_total += weight
        weighted_sum += weight * bond.duration_years
    if weight_total <= 0:
        return None
    return weighted_sum / weight_total

File: backend/internal/domain/portfolio/duration_metrics.go
package portfolio

import (
	"github.com/tonatos/bond-monitor/backend/internal/domain/bonds"
)

// RateSensitiveDuration returns duration for rate-risk (floaters ≈ policy floater years).
func RateSensitiveDuration(b bonds.BondRecord, dp DurationPolicy) *float64 {
	if b.IsFloatingCoupon() {
		v := dp.FloaterRateDurationYears
		return &v
	}
	return b.DurationYears()
}

// DurationHolding is weighted by market value for duration metrics.
type DurationHolding struct {
	ISIN            string
	MarketValueRub  *float64
}

func WeightedDurationByPurchase(
	positions []PortfolioPosition,
	universeByISIN map[string]bonds.BondRecord,
	durationPolicy DurationPolicy,
) *float64 {
	var weightTotal, weightedSum float64
	for _, pos := range positions {
		bond, ok := universeByISIN[pos.ISIN]
		if !ok {
			continue
		}
		duration := RateSensitiveDuration(bond, durationPolicy)
		if duration == nil {
			continue
		}
		weight := pos.PurchaseAmountRub
		weightTotal += weight
		weightedSum += weight * *duration
	}
	if weightTotal <= 0 {
		return nil
	}
	v := weightedSum / weightTotal
	return &v
}

func WeightedDurationByMarket(
	holdings []DurationHolding,
	universeByISIN map[string]bonds.BondRecord,
	durationPolicy DurationPolicy,
) *float64 {
	var weightTotal, weightedSum float64
	for _, holding := range holdings {
		bond, ok := universeByISIN[holding.ISIN]
		if !ok {
			continue
		}
		duration := RateSensitiveDuration(bond, durationPolicy)
		if duration == nil {
			continue
		}
		if holding.MarketValueRub == nil || *holding.MarketValueRub <= 0 {
			continue
		}
		weight := *holding.MarketValueRub
		weightTotal += weight
		weightedSum += weight * *duration
	}
	if weightTotal <= 0 {
		return nil
	}
	v := weightedSum / weightTotal
	return &v
}
